experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -71,6 +82,193 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+
+        if (isBackedge(result, user))
+          return condBrOp;
+      }
+    }
+  }
+  funcOp->emitError() << "Could not find backedge within BB " << bb << "\n";
+  return nullptr;
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceSaveCommits() {
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+
+  // Don't do anything if there are no SaveCommits to place
+  if (placements.getPlacements<handshake::SpecSaveCommitOp>().empty())
+    return success();
+
+  // The save commits are a result of a control branch being in the BB
+  // The control path for the SC needs to replicate the branch
+  unsigned bb = getLogicBB(specOp).value();
+  ConditionalBranchOp controlBranch = findControlBranch(specOp, bb);
+  if (controlBranch == nullptr)
+    return failure();
+
+  // To connect a Save-Commit, two control signals are sent from the Speculator
+  // and are merged before reaching the Save-Commit.
+  // The tokens take differents paths. One needs to always reach the SC,
+  // the other should follow the actual branches similarly to the Commits
+  builder.setInsertionPointAfterValue(specOp.getSCCommitCtrl());
+  auto branchDiscardNonSpec = builder.create<handshake::SpeculatingBranchOp>(
+      controlBranch.getLoc(), controlBranch.getTrueResult() /* spec tag */,
+      controlBranch.getConditionOperand());
+  inheritBB(specOp, branchDiscardNonSpec);
+
+  // This branch will propagate the signal SCCommitControl according to
+  // the control branch condition, which comes from branchDiscardNonSpec
+  auto branchDiscardCond = builder.create<handshake::ConditionalBranchOp>(
+      branchDiscardNonSpec.getLoc(), branchDiscardNonSpec.getTrueResult(),
+      specOp.getSCCommitCtrl());
+  inheritBB(specOp, branchDiscardCond);
+
+  // Create a conditional branch driven by SCBranchControl from speculator
+  auto branchSaveCommitCond = builder.create<handshake::ConditionalBranchOp>(
+      branchDiscardCond.getLoc(), specOp.getSCBranchCtrl(),
+      branchDiscardCond.getTrueResult());
+  inheritBB(specOp, branchSaveCommitCond);
+
+  // We create a Merge operation to join SCCSaveCtrl and SCCommitCtrl signals
+  SmallVector<Value, 2> mergeOperands;
+  mergeOperands.push_back(specOp.getSCSaveCtrl());
+
+  // We need to send the control token to the same path that the speculative
+  // token followed. Hence, if any branch output leads to a backedge, replicate
+  // the branch in the SaveCommit control path.
+  auto isBranchBackedge = [&](Value result) {
+    return llvm::any_of(result.getUsers(), [&](Operation *user) {
+      return isBackedge(result, user);
+    });
+  };
+
+  // Check if trueResult of controlBranch leads to a backedge (loop)
+  if (isBranchBackedge(controlBranch.getTrueResult()))
+    mergeOperands.push_back(branchSaveCommitCond.getTrueResult());
+
+  // Check if falseResult of controlBranch leads to a backedge (loop)
+  if (isBranchBackedge(controlBranch.getFalseResult()))
```

Comment:
I would change this to if(is true back edge) {...} else if(is false back edge) {...} else { return failed; }, the pass should fail if we can't find the backedge for the control token:
even if this is checked earlier, for anyone reading it should be clear that it has to be either one path or the other.

Reply:
Changed to your suggestion.